src/tree/mod.rs

```diff
@@ -737,121 +750,83 @@ impl RatchetTree {
+                        continue;
+                    }
+                } else {
+                    // Then verify the right child
+                    if let Some(right_parent_hash_field) = self.nodes[right].parent_hash() {
+                        if right_parent_hash_field == current_hash_left {
+                            continue;
                         }
-                        let left_option = &nodes[left_index.as_usize()];
-                        let right_option = &nodes[right_index.as_usize()];
-                        let own_hash = node.hash(ciphersuite).unwrap();
-                        if let Some(right) = right_option {
-                            if let Some(left) = left_option {
-                                let left_parent_hash = left.parent_hash().unwrap_or_else(Vec::new);
-                                let right_parent_hash =
-                                    right.parent_hash().unwrap_or_else(Vec::new);
-                                if (left_parent_hash != own_hash) && (right_parent_hash != own_hash)
-                                {
-                                    return false;
-                                }
-                                if left_parent_hash == right_parent_hash {
-                                    return false;
-                                }
-                            } else if right.parent_hash().unwrap() != own_hash {
-                                return false;
-                            }
-                        } else if let Some(left) = left_option {
-                            if left.parent_hash().unwrap() != own_hash {
-                                return false;
+                    } else {
+                        // Iterate over children until we find a node that is neither blank
+                        // nor a leaf node
+                        let mut child = right;
+                        while !self.nodes[child].is_full_parent() {
+                            let right_option = treemath::left(child);
+                            match right_option {
+                                // Keep iterating
+                                Ok(r) => child = r,
+                                // We didn't find any full parent node in this path
+                                Err(_) => return false,
                             }
                         }
-                    }
-                    NodeType::Leaf => {
-                        if let Some(kp) = &node.key_package {
-                            if i % 2 != 0 {
-                                return false;
-                            }
-                            if kp.verify().is_err() {
-                                return false;
-                            }
+                        // Unwrapping here is safe, because we know it is a full parent node
+                        if self.nodes[child].parent_hash().unwrap() == current_hash_left {
```

Comment:
We know it's not a leaf, but it could still be a blank, so we can't `unwrap` here, I think.